reference-architecture/azure-ansible/3.6/allinone.sh

```diff
@@ -431,6 +431,213 @@ cat <<EOF > /home/${AUSERNAME}/postinstall.yml
+  - name: Stage 9 - OCCREATE SECRET
+    command: "oc create secret generic sso-app-secret --from-file={{idm_dir}}/jgroups.jceks --from-file={{idm_dir}}/sso-https.jks --from-file={{idm_dir}}/truststore.jks"
+  - name: Stage 10 - OCCREATE SECRET ADD
+    command: "oc secret add sa/{{sso_project}}-service-account secret/sso-app-secret"
+  - name: Stage 11 - Create App Parameters
+    blockinfile:
+       path: "{{idm_dir}}/sso.params"
+       create: yes
+       block: |
+         HOSTNAME_HTTP="nlogin.{{sso_domain}}"
+         HOSTNAME_HTTPS="login.{{sso_domain}}"
+         APPLICATION_NAME="{{sso_project}}"
+         HTTPS_KEYSTORE="sso-https.jks"
+         HTTPS_PASSWORD="{{idm_xpassword}}"
+         HTTPS_SECRET="sso-app-secret"
+         JGROUPS_ENCRYPT_KEYSTORE="jgroups.jceks"
+         JGROUPS_ENCRYPT_PASSWORD="{{idm_xpassword}}"
+         JGROUPS_ENCRYPT_SECRET="sso-app-secret"
+         SERVICE_ACCOUNT_NAME={{sso_project}}-service-account
+         SSO_REALM=cloud
+         SSO_SERVICE_USERNAME="{{sso_username}}"
+         SSO_SERVICE_PASSWORD="{{sso_password}}"
+         SSO_ADMIN_USERNAME=admin
+         SSO_ADMIN_PASSWORD="{{sso_password}}"
+         SSO_TRUSTSTORE=truststore.jks
+         SSO_TRUSTSTORE_PASSWORD="{{idm_xpassword}}"
+
+  - name: Stage 10 - OCCREATE SECRET ADD
+    command: oc new-app sso71-postgresql --param-file {{idm_dir}}/sso.params -l app=sso71-postgresql -l application=sso -l template=sso71-https
+  - set_fact: sso_token_url="https://login.{{sso_domain}}/auth/realms/cloud/protocol/openid-connect/token"
+  - name: Pause for app create
+    pause:
+      minutes: 4
+  - name: Login to SSO and Get Token
+    uri:
+      url: "{{sso_token_url}}"
+      method: POST
+      body: "grant_type=password&client_id=admin-cli&username={{sso_username}}&password={{sso_password}}"
+      return_content: yes
+      status_code: 200
+      validate_certs: no
+    register: login
+    until: login.status == 200
+    retries: 90
+    delay: 30
+  - debug: var=login.json.access_token
+  - name: Create SSO Client for Openshift
+    uri:
+      url: "https://login.{{sso_domain}}/auth/realms/cloud/clients-registrations/default"
+      method: POST
+      headers:
+           "Authorization": "bearer {{login.json.access_token}}"
+           "Content-Type": "application/json"
+      body: "{{ create_data | to_json }}"
+      return_content: yes
+      status_code: 201
+      validate_certs: no
+    register: create
+  - debug: var=create.json.secret
+- hosts: masters
+  vars_files:
+    - ssovars.yml
+  tasks:
+  - name: Copy xpass.conf to masters
+    copy:
+      src:  "{{idm_dir}}/xpaas.crt"
+      dest: /etc/origin/master/xpaas.crt
+      owner: root
+      mode: 0600
+  - name: Setup SSO Config
+    blockinfile:
+      backup: yes
+      dest: /etc/origin/master/master-config.yaml
+      insertafter: HTPasswdPasswordIdentityProvider
```

Comment:
OCP allows having multiple identity providers. In this particular case it will be htpasswd and sso, is that on purpose? Otherwise I'll remove the htpasswd.

Reply:
Just in case, as its chicken and egg problem, if sso does deploy you lose the ability to login.

Reply:
Because all the rest of the providers can be used inside of sso. So sso gives you a gui for adding providers and controlling them.